Fleet API support

tomhollander opened the pull request:
Updated the library to support the Fleet API.
Removed the code that supported the pure API (browserless) authentication since it doesn't work anymore.
Updated WPF sample
Added new ASP.NET Core web sample 
Updated README.

Note, I've tested this with my own apps registered for the Tesla Fleet API. While I can successfully obtain tokens, I'm yet to figure out the partner registration process so I haven't proven the tokens can be used. I'll get to that soon but I was keen to get others to try the API in the meantime.

tomhollander commented:
@brianflex @briangru @ramonsmits I'd appreciate your review if you have the time.

ramonsmits commented on library/TeslaAuthHelper.cs:
```diff
@@ -167,137 +160,40 @@ public async Task<Tokens> RefreshTokenAsync(string refreshToken, CancellationTok
 
         async Task<Tokens> ExchangeCodeForBearerTokenAsync(string code, HttpClient client, CancellationToken cancellationToken)
         {
             var body = new JObject
```
Maybe use [System.Text.Json.Nodes](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.nodes?view=net-6.0) ? Could potentially remove the dependency on Newtonsoft.Json

tomhollander replied:
Thanks will do

ramonsmits commented on library/TeslaAuthHelper.cs:
```diff
@@ -167,137 +160,40 @@ public async Task<Tokens> RefreshTokenAsync(string refreshToken, CancellationTok
-            };
+                {"redirect_uri", redirectUri},
+                { "scope", scopes },
+                { "audience", "https://fleet-api.prd.na.vn.cloud.tesla.com" }
```
Not sure, but isn't this url based on region? See https://developer.tesla.com/docs/fleet-api#overview

- North America, Asia-Pacific (excluding China): https://fleet-api.prd.na.vn.cloud.tesla.com/
- Europe, Middle East, Africa: https://fleet-api.prd.eu.vn.cloud.tesla.com/

tomhollander replied:
Ah good catch this was on my Todo list and I missed it. Although it would be good to test if this actually matters.

ramonsmits reviewed (commented):
I only quickly reviewed the lib changes. I'll review but more importantly test from the EU (Netherlands).

tomhollander commented on library/Scopes.cs:
```diff
@@ -0,0 +1,29 @@
+        public static string EnergyDeviceData = "energy_device_data";
+        public static string EnergyCommands = "energy_cmds";
+
+        public static string GetScopeString(params string[] scopes)
```
Does that still work with `params`?

ramonsmits replied:
No, it doesn't but it would allow to just pass for example a HashSet or List.

tomhollander commented on library/Scopes.cs:
```diff
@@ -0,0 +1,29 @@
+        public static string GetScopeString(params string[] scopes)
+        {
+            var sb = new StringBuilder();
+            sb.Append("openid offline_access ");
```
From my reading of the docs, these are always needed, so I thought it was best if users don't need to specify them every time.

tomhollander commented on library/TeslaAuthHelper.cs:
```diff
@@ -33,15 +33,29 @@ public class TeslaAuthHelper
         const string TESLA_CLIENT_ID = "81527cff06843c8634fdc09e8ac0abefb46ac849f38fe1e431c2ef2106796384";
         const string TESLA_CLIENT_SECRET = "c7257eb71a564034f9419ee651c7d0e5f7aa6bfbd18bafb5c5c033b093bb2fa3";
+        const string TESLA_REDIRECT_URI = "https://auth.tesla.com/void/callback\"";
+        const string TESLA_SCOPES = "openid email offline_access";
```
These are the legacy Owner API scopes.

tomhollander commented:
Thanks @ramonsmits - I addressed most of your feedback, and also improved comments and added a new console sample.
I didn't move away from Newtonsoft.Json as that would require bumping the .NET Standard version to 2.1 which would change the set of supported platforms. Not sure if this matters or not but I didn't want to do it without some thought. 
LMK what you think of the changes.

ramonsmits commented on library/TeslaAuthHelper.cs:
```diff
@@ -368,115 +253,26 @@ static string GetBaseAddressForRegion(TeslaAccountRegion region)
                     throw new NotImplementedException("Fell threw switch in GetBaseAddressForRegion for " + region);
```
```suggestion
                    throw new NotSupportedException("Region not supported: " + region);

```

ramonsmits commented on library/TeslaAuthHelper.cs:
```diff
@@ -368,115 +253,26 @@ static string GetBaseAddressForRegion(TeslaAccountRegion region)
+                case TeslaAccountRegion.Europe:
+                    return "https://fleet-api.prd.eu.vn.cloud.tesla.com";
+                default:
+                    throw new NotImplementedException("Fell threw switch in GetAudienceAdressForRegion for " + region);
```
```suggestion
                    throw new NotSupportedException("Region not supported: " + region);

```

tomhollander commented:
@ramonsmits Thanks for the fast reviews. Have you been able to test it yet?

ramonsmits commented:
I checked but it seems I need to create an account. I don't know if this
will affect my regular account as it is only free for a short time so
unfortunately not going to test to find out.

On Tue, Oct 17, 2023, 10:28 Tom Hollander ***@***.***> wrote:

> @ramonsmits <https://github.com/ramonsmits> Thanks for the fast reviews.
> Have you been able to test it yet?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/tomhollander/TeslaAuth/pull/36#issuecomment-1765926094>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABFLJRGTXCTJPI3BE24DI3X7Y6SRAVCNFSM6AAAAAA6B4FSZ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTONRVHEZDMMBZGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
